features/nfc/source/nfc/NFCController.cpp

```diff
@@ -0,0 +1,220 @@
+
+    // rf_protocols.initiator_t1t = initiator_tech.nfc_type1;
+    // rf_protocols.initiator_t2t = initiator_tech.nfc_type2;
+    // rf_protocols.initiator_t3t = initiator_tech.nfc_type3;
+    // rf_protocols.initiator_iso_dep = initiator_tech.nfc_iso_dep_a || initiator_tech.nfc_iso_dep_b;
+    // rf_protocols.initiator_nfc_dep = initiator_tech.nfc_nfc_dep_a || initiator_tech.nfc_nfc_dep_f_212 || initiator_tech.nfc_nfc_dep_f_424;
+    
+    // rf_protocols.target_t1t = target_tech.nfc_type1;
+    // rf_protocols.target_t2t = target_tech.nfc_type2;
+    // rf_protocols.target_t3t = target_tech.nfc_type3;
+    rf_protocols.target_iso_dep = target_tech.nfc_iso_dep_a || target_tech.nfc_iso_dep_b;
+    // rf_protocols.target_nfc_dep = target_tech.nfc_nfc_dep_a || target_tech.nfc_nfc_dep_f_212 || target_tech.nfc_nfc_dep_f_424;
+    
+    return rf_protocols;
+}
+
+nfc_err_t NFCController::configure_rf_protocols(nfc_rf_protocols_bitmask_t rf_protocols)
+{
+    if( _discovery_running ) {
+        // Cannot configure RF protocols if discovery is running
+        return NFC_ERR_BUSY;
+    }
+
+    // Map to NFC techs
+    nfc_tech_t initiator_tech = {0};
+    nfc_tech_t target_tech = {0};
+
+    // Note: we only support ISO-DEP tag emulation in this release,
+    // so mask out all other protocols
+
+    target_tech.nfc_iso_dep_a = target_tech.nfc_iso_dep_b =  true;
+
+    // Configure polling options (no need to set bailing flags as we're only using target mode)
+    polling_options_t options = {0};
+
+    transceiver_set_protocols(_transceiver, initiator_tech, target_tech, options);
+}
+
+nfc_err_t NFCController::start_discovery()
+{
+    if( _discovery_running ) {
+        // Cannot start discovery if it's already running
+        return NFC_ERR_BUSY;
+    }
+
+    transceiver_poll(_transceiver, &NFCController::s_polling_callback, this /* use this as callback argument */);
+}
+
+nfc_err_t NFCController::cancel_discovery()
+{
+    if( !_discovery_running ) {
+        return NFC_OK;
+    }
+
+    transceiver_abort(_transceiver);
+}
+
+nfc_transceiver_t* NFCController::transceiver() const
+{
+    return _transceiver;
+}
+
+void NFCController::polling_callback(nfc_err_t ret)
+{
+    // Polling has completed
+    _discovery_running = false;
+
+    if( ret == NFC_OK ) {
+        // Check if a remote initiator was detected and if so, instantiate it
+        if( !transceiver_is_initiator_mode(_transceiver) ) {        
+            nfc_tech_t active_tech = transceiver_get_active_techs(_transceiver);
+            if( active_tech.nfc_iso_dep_a || active_tech.nfc_iso_dep_b ) {
+                Type4RemoteInitiator type4_remote_initiator_ptr = new (std::nothrow) Type4RemoteInitiator(_transceiver, _ndef_buffer, _ndef_buffer_sz);
+                if( type4_remote_initiator_ptr == NULL ) {
+                    // No memory :(
+                    SharedPtr<Type4RemoteInitiator> type4_remote_initiator( type4_remote_initiator_ptr );
```

Comment:
Is calling `_delegate->on_nfc_initiator_discovered` with a `NULL` shared pointer intended ?